cli/cli.js

```diff
@@ -72,38 +76,124 @@ const runYoCommand = (cmd, args, options, opts) => {
+        command.option(opt.option, opt.desc, opt.default);
+    });
+
+    if (!opts.cliOnly) {
+        const namespace = opts.blueprint ? `${packageNameToNamespace(opts.blueprint)}:${key}` : `${JHIPSTER_NS}:${key}`;
+        const generator = env.create(namespace, { options: { help: true } });
+        Object.entries(generator._options).forEach(([key, value]) => {
+            if (value.hide || key === 'help') {
+                return;
+            }
+            let cmdString = '';
+            if (value.alias) {
+                cmdString = `-${value.alias}, `;
+            }
+            cmdString = `${cmdString}--${key}`;
+            if (value.type === String) {
+                cmdString = `${cmdString} <value>`;
+            }
+            command.option(cmdString, value.description, value.default);
+        });
+    }
+
     command
         .allowUnknownOption()
         .description(opts.desc)
-        .action(args => {
-            const options = getCommandOptions(packageJson, process.argv.slice(2));
+        .action((first, second, third) => {
+            let args;
+            let cmdOptions;
+            let unknownArgs;
+            logger.debug(`first command arg: ${toString(first)}`);
+            logger.debug(`second command arg: ${toString(second)}`);
+            logger.debug(`third command arg: ${toString(third)}`);
+
+            if (opts.argument) {
+                // Option that contains arguments
+                // first=arguments second=cmdOptions third=unknownArgs
+                if (Array.isArray(first)) {
+                    // Var args unknown options are concatenated.
+                    // consider valid args before first unknown option (starts with -).
+                    args = [];
+                    unknownArgs = [];
+                    let unknown = false;
+                    first.forEach(item => {
+                        if (item.startsWith('-')) {
+                            unknown = true;
+                        }
+                        if (unknown) {
+                            unknownArgs.push(item);
+                        } else {
+                            args.push(item);
+                        }
+                    });
+                } else if (first !== undefined) {
+                    args = [first];
+                }
+                cmdOptions = second;
+                if (third) {
+                    unknownArgs = (unknownArgs || []).concat(third);
+                }
+            } else {
+                // Option that doesn't contain arguments
+                // first=cmdOptions second=unknownArgs
+                args = [];
+                cmdOptions = first;
+                unknownArgs = second || [];
+            }
+
+            const firstUnknownArg = Array.isArray(unknownArgs) && unknownArgs.length > 0 ? unknownArgs[0] : undefined;
+            if (key === 'app' && firstUnknownArg !== undefined && !firstUnknownArg.startsWith('-')) {
```

Comment:
why is unknown command handling done for `app` command, we don't have second level commands also I didn't see the didYouMean part working here, can you double check. The didYouMean hint should be for global unknown command like `jhipster entittty`

Reply:
`If you use .command('*') to add a default command, you may be be able to use isDefault:true with a normal command instead.`
Commander 5 changed the behaviour and the `*` event is only emitted if there is no default command.
So `jhipster entittty` is treated like `jhipster app entittty`, since `app` doesn't takes any argument, we are good for didYouMean.
It makes sense to treat `jhipster foo.jdl` as `jhipster app foo.jdl` since app is the default command.

For jhipster 7 we could remove `.allowUnknownOption()`.
This will require every command and option to be registered, otherwise it will throw an exception.

Reply:
Can I close this review @deepu105 ?

Reply:
if you are getting did you mean suggestions then yes, it wasn't working for me. May be remove `allowUnknownOption` as its not a breaking change that would affect anyone

Reply:
I didn't realised it wasn't working. I thought it was a design preference.
Fixed now.

Reply:
About allowUnknownOption, we must be sure to cover blueprint's custom commands support.